socialhome/content/viewsets.py

```diff
@@ -47,31 +48,48 @@ class ContentViewSet(mixins.CreateModelMixin, mixins.RetrieveModelMixin, mixins.
         `visibility` should be one of: `public`, `site`, `limited`, `self`.
 
     share:
-        Share content
+        Content sharing
 
-        No additional required values. Will share the content using the authenticated profile. Returns
-        share content ID as `content_id`.
+        No additional required values. Create or remove a share of this content.
+
+        Successful create share returns content ID as `content_id`.
     """
     queryset = Content.objects.none()
     serializer_class = ContentSerializer
     permission_classes = (IsOwnContentOrReadOnly,)
     throttle_classes = (CreateContentThrottle,)
 
-    def get_queryset(self):
-        if self.request.user.is_staff:
-            return Content.objects.all()
-        return Content.objects.visible_for_user(self.request.user)
+    def _share(self):
+        content = self.get_object()
+        try:
+            share = content.share(self.request.user.profile)
+        except ValidationError as e:
+            raise exceptions.ValidationError(e.message)
```

Comment:
Can't you just do 

```python
except ValidationError:
    raise
```

?

Reply:
We don't want to raise *Django* `ValidationError`, we want to raise *Django REST framework* `ValidationError`. If you look at it closely, they're not the same class. The reason is that if we just pass Django validation error down from the REST viewset, it wont give the same nice error message to the client as the Django REST validation error. That is why we "convert" it here.

Reply:
Ah, right.